### ISSUE TITLE
Slow `compact_allocate`

### ISSUE THREAD

**Issue by aalekseyev** [ocaml/ocaml#12264]
I'm seeing what I think is a pathological performance in `compact_allocate`.

The following program allocates ~1GB of memory and does compaction:

```
let () = Gc.set { (Gc.get ()) with minor_heap_size = 1048576 }
let v = List.init (1_000_000_000 / (8 * 3)) (fun i -> i)
let () = Gc.compact ()
let (_ : _) = Sys.opaque_identity v
```

The compaction takes ~5s, almost all of the time taken by `compact_allocate`.
Unfortunately, this benchmark is fragile: in particular I can't reproduce the issue without the [Gc.set] call, even though it doesn't matter what I set the minor heap size to. (it only matters that it's different from the initial value).

Here's a perf report (~200 samples in total here):

```
       │    00000000005a3f30 <compact_allocate>:
       │    compact_allocate():
       │      mov    compact_fl,%rax
       │      mov    -0x30(%rax),%rcx
       │      mov    -0x28(%rax),%rdx
       │      sub    %rcx,%rdx
       │      cmp    $0xf,%rdx
       │    ↓ ja     3c
       │      nop
       │20:   mov    -0x20(%rax),%rax
       │      mov    -0x30(%rax),%rcx
       │      mov    -0x28(%rax),%rdx
       │      sub    %rcx,%rdx
       │      cmp    $0xf,%rdx
       │    ↑ jbe    20
       │      mov    %rax,compact_fl
       │3c:   cmp    %rdx,%rdi
       │    ↓ jbe    5c
       │      nop
       │48:┌─→mov    -0x20(%rax),%rax
       │   │  mov    -0x30(%rax),%rcx
 91.70 │   │  mov    -0x28(%rax),%rdx
       │   │  sub    %rcx,%rdx
       │   │  cmp    %rdi,%rdx
  8.30 │   └──jb     48
       │5c:   add    %rcx,%rdi
       │      mov    %rdi,-0x30(%rax)
       │      add    %rcx,%rax
       │    ← retq
```

We looked at this with Leo (@lpw25) and our conjecture is that the allocation algorithm suffers from the known first-fit performance problem: for every element of the list, the allocation algorithm scans all blocks on the free list. In our case we have ~45 blocks, so allocation is slower than what it should be by that factor.

One improvement idea, suggested by Leo, is to implement next-fit instead of first-fit. 
If we're worried about sacrificing even the smallest bit of compactness, we can implement a tweak to the next-fit algorithm: maintain a pointer in the free list which tells us which blocks have been used already, and which ones are completely free (therefore subject to being dropped). Then we make sure to avoid touching a completely free block unless we've scanned through all blocks that have been used.

I can also think of a more efficient (in the worst-case) implementation of first-fit semantics, where instead of scanning the free list from the beginning every time, we scan it backwards from the point where the previous scan ended (we need to keep track of extra info for this: the monotonically increasing subsequence of the prefix). However, next-fit seems just simpler.

**Comment by avsm** [comment]
Thanks for the report; which version of the compiler is this against?  I assume 4.14.x.  Might you be able to try out 5.0 with #12193 (the compactor) to see if it's similarly affected, or are you looking for a solution against 4.14-dev only?

**Comment by aalekseyev** [comment]
Sorry, yes, this is in 4.14.

**Comment by aalekseyev** [comment]
Given that the benchmark is pretty fragile, the easiest way to answer if something is affected is by code inspection (to see if it uses first-fit allocation).

I can't even see `compact.c` in that PR, so I'm a bit lost. If anyone knows where the corresponding code in OCaml 5 is (and whether that's likely to be affected), that'd be helpful.

**Comment by lpw25** [comment]
I think the part of the heap that is compacted in 5 is size segmented, so shouldn't have this problem. I'm unclear on whether we intend to do bug fix releases for 4.14 still. If so it's probably pretty easy to fix this one. (Also, we'd like to port that fix onto our version of the compiler since it is still 4.14 based).

**Comment by damiendoligez** [comment]
I'd be interested in more details about the program that triggers this bug. Does it, like the repro case, allocate only one size of blocks? Are they also of size 2?

I don't think we should worry about extracting the last tiny bit of compactness, and this gives us some room for a simple fix.

One solution is to drop a free block as soon as it's too small to fit the current request. Fast and simple (simpler even than the current code) but it can leave holes of arbitrary size.

A slightly more complex solution is to have two pointers into the list, one for small blocks and one for large blocks, and likewise advance each pointer as soon as the current request doesn't fit. This guarantees holes smaller than the cut-off size while still guaranteeing constant-time allocations.

I'll do a PR against 4.14.

**Comment by aalekseyev** [comment]
The program where I saw the bug was simply a benchmark of something unrelated. The benchmark was literally allocating a long list of ints at startup. So yeah, it was pretty much identical.

**Comment by lpw25** [comment]
I should say that we don't think this is a particularly important bug, there is some small possibility that it's slowing down a lot of compactions and no one is noticing because compaction is already really slow, but it also might just affect artificial programs like the benchmark we saw it on. I suggested sending it here because it is a pathological behavior and because it looks very easy to fix. If it is indeed easy to fix then it's probably worth fixing, but if it starts to look even slightly difficult then it's probably not worth the effort to fix since at some point everyone will be on OCaml 5 and it won't matter.
